File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        
        int total = 0;
        int time = 0;
        int idx = 0;
        Queue<Integer> bridge = new LinkedList<>();
        while (idx < truck_weights.length || total > 0) {
            time++;
            
            if(bridge.size() == bridge_length) {
                total -= bridge.poll();
            }
            
            if (idx<truck_weights.length && total + truck_weights[idx] <= weight) {
                bridge.add(truck_weights[idx]);
                total += truck_weights[idx];
                idx++;
            } else {
                bridge.add(0);
            }
        }
        
        return time;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.swift
import Foundation

func solution(_ bridge_length:Int, _ weight:Int, _ truck_weights:[Int]) -> Int {
    var time = 0
    
    var waitHead = 0
    var bridgeHead = 0
    var bridge = [(Int,Int)]() // weight time
    var total = 0
    while(waitHead < truck_weights.count || bridgeHead  < bridge.count) {
        // 현재 시간 계산해서 내보내기
        while bridgeHead < bridge.count {
            let (w,t) = bridge[bridgeHead]
            if time - t >= bridge_length {
                total -= w
                bridgeHead += 1
            } else {
                break
            }
        }
        
        // 들어갈수있는 트럭 넣기
        if(waitHead < truck_weights.count) {
            let nxt = truck_weights[waitHead]
            if (total + nxt <= weight) { // 다음꺼 들어갈수있음
                bridge.append((nxt, time))
                waitHead += 1
                total += nxt
            }
        }
        
        time += 1
    }
    
    return time
}